Services/DataBaseService/Utils/MigrationEngine.cs

```diff
@@ -0,0 +1,269 @@
+                Console.WriteLine(e.Message);
+                throw;
+            }
+        }
+
+        private void WriteDownExecutedScripts(string connectionString, 
+            Dictionary<string, ExecutedScript> scriptsToWriteDown)
+        {
+            var insertRow = new StringBuilder();
+            insertRow.AppendLine("INSERT INTO [dbo].[ExecutedScripts] ");
+            insertRow.AppendLine("(FileName, ExecutionTime, Code) ");
+            insertRow.AppendLine("VALUES (@fileName, @executionTime, @code);");
+            string lastScriptToWrite = "[no script written]";
+
+            try
+            {
+                using (var conn = new SqlConnection(connectionString))
+                {
+                    conn.Open();
+                    SqlTransaction writingTransaction = conn.BeginTransaction();
+                    try
+                    {
+                        foreach (var script in scriptsToWriteDown)
+                            using (SqlCommand command = new SqlCommand(insertRow.ToString(), conn, writingTransaction))
+                            {
+                                lastScriptToWrite = script.Key;
+                                command.Parameters.AddWithValue("@fileName", script.Key);
+                                command.Parameters.AddWithValue("@executionTime", script.Value.ExecutionTime);
+                                command.Parameters.AddWithValue("@code", script.Value.ExecutedCode);
+                                command.ExecuteNonQuery();
+                            }
+                        writingTransaction.Commit();
+                    }
+                    catch (SqlException e)
+                    {
+                        // TODO replace with logs
+                        Console.WriteLine(e.Message +
+                            $"\n\tWarning! The {lastScriptToWrite} script cannot be marked as executed.");
+                        try
+                        {
+                            writingTransaction.Rollback();
+                            // TODO replace with logs
+                            Console.WriteLine("\tWriting rollback is successful.");
+                        }
+                        catch
+                        {
+                            // TODO replace with logs
+                            Console.WriteLine("\tCouldn't rollback writing.");
+                        }
+                        throw e;
+                    }
+                    finally
+                    {
+                        writingTransaction.Dispose();
+                    }
+                }
+            }
+            catch (SqlException e) when (e.Number == -2)
+            {
+                // TODO replace with logs
+                Console.WriteLine(e.Message + "\n\tCouldn't write down executed scripts: the connection timeout.");
+                throw e;
+            }
+            catch (Exception e) when (!(e is SqlException))
```

Comment:
May be I'm wrong but do we really need that `(!(e is SqlException))` check?

Reply:
We're handling SqlExceptions (to be honest, only logging and throwing again, but does it matter?) inside the inner try-catch, so I suppose there's no need to catch them again. However, I may be wrong too